docs/core/settings/variables.md

```diff
@@ -27,24 +30,118 @@ You can use special variables in several places:
 * [[link,auth_ldap]], [[link,auth_sql]], and [[link,userdb]] query strings
 * Log prefix for imap/pop3 process
 
-## Global Variables
-
-Global variables that work everywhere are:
+## Variable expansion syntax
+
+[[changed,var_expand]]
+
+We have introduced an entirely new variable expansion syntax.
+See [[link,var_expand]] for in-depth details how it works.
+
+The basic syntax is `%{variable (| filter | filter ...)}`, which means that most existing
+variables work, but there are some changes, so check variable usage carefully when converting old syntax.
+
+The simple case of just getting a value of variable is `%{variable}`.
+These can be in middle of strings.
+
+Another syntax is `%{provider:variable}`, where the value is provided by
+a provider. There are global providers, and context-specific providers.
+
+A variable can be then filtered with various filters, such as `%{variable | upper}` to get
+uppercase representation of variable. You can chain as many filters as you need.
+
+Filters can accept parameters, both positional and named. E.g. `%{literal('\r\n\')}` will expand
+to CR LF. `%{user | substr(1)}` will take first character of username. Example of named parameters
+would be `%{user | md5(rounds=1000,salt='pepper')}`
+
+You can use `%%{variable}` to escape this, and emit `%{variable}`.
+
+Filters accept strings, numbers and variables as parameters. Parameters can be positional or
+named key-value pairs. Key names cannot be variables.
+The left side of pipe character (`|`) is provided as input to a filter. Some filters can be used
+in place of variables, e.g. lookup, literal and if.
+
+When value is missing or empty, you can use the `default` filter to provide value. Missing variables
+will cause errors and must be negated with default. This does not apply to all providers, some
+providers return empty when value is missing.
+
+If the last filter would output binary data, the data is encoded with `hexlify` filter by default.
+To avoid this, you can use `text` filter, which will sanitize the input and mark it as text.
+
+The new syntax also supports simple maths, you can do one operation. E.g. `%{port + 1000}`.
+Addition, substraction, multiplication, division and modulo operations are supported for now.
+
+A special case for modulo operation is that it can be applied to binary input, e.g. `md5 % 256`.
+The input is treated as 64-bit unsigned number and modulo is taken from that.
+
+All strings must be encapsulated with `"` or `'`, and you can escape them using `\\` within string.
+Numbers when used as parameters must be provided without quotes.
+
+## List of filters
+
+All parameters are strings unless stated otherwise.
+If parameter is `any`, it accepts both numbers and strings.
```

Comment:
and bytes?

Reply:
there is no special bytes type for variables, variables are either numbers or strings.